File: Dynamics-iOS-Objective-C/GoodBaseline/HttpViewController.m
/* Copyright (c) 2021 BlackBerry Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

#import "HttpViewController.h"

@interface HttpViewController ()

@property (weak, nonatomic) IBOutlet UITextField *urlText;
@property (weak, nonatomic) IBOutlet UILabel *statusLabel;
@property (weak, nonatomic) IBOutlet UIButton *getURLButton;
@property (weak, nonatomic) IBOutlet UITextView *dataReceivedTextView;

- (IBAction)clearDataView:(id)sender;
- (IBAction)connectToURL:(id)sender;
- (NSURL *)smartURLForString:(NSString *)str;
- (void)receiveDidStart;

@end

@implementation HttpViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    // Do any additional setup after loading the view.

    self.urlText.text = @"https://developer.blackberry.com";

}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

/*
#pragma mark - Navigation

// In a storyboard-based application, you will often want to do a little preparation before navigation
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    // Get the new view controller using [segue destinationViewController].
    // Pass the selected object to the new view controller.
}
*/


#pragma mark NSURLConnection Delegate Methods

- (void)connection:(NSURLConnection *)connection didReceiveResponse:(NSURLResponse *)response {
    // A response has been received, this is where we initialize the instance var you created
    // so that we can append data to it in the didReceiveData method
    // Furthermore, this method is called each time there is a redirect so reinitializing it
    // also serves to clear it
    
    dataBuffer = [[NSMutableData alloc] init];
}

- (void)connection:(NSURLConnection *)connection didReceiveData:(NSData *)data {
    // Append the new data to the instance variable you declared
    
    [dataBuffer appendData:data];
}

- (NSCachedURLResponse *)connection:(NSURLConnection *)connection
                  willCacheResponse:(NSCachedURLResponse*)cachedResponse {
    // Return nil to indicate not necessary to store a cached response for this connection
    return nil;
}

- (void)connectionDidFinishLoading:(NSURLConnection *)connection {
    // The request is complete and data has been received
    NSString *receivedData = [[NSString alloc] initWithData:dataBuffer encoding:NSUTF8StringEncoding];
    
    self.dataReceivedTextView.text = receivedData;
    self.statusLabel.text = @"Data loaded.";
    self.getURLButton.enabled = FALSE;
    
    // reset the buffer
    [dataBuffer setLength:0];
    
    // set currentConnection to nil to avoid problems if aborting after the connection has gone away
    connection = nil;
    
}

- (void)connection:(NSURLConnection *)connection didFailWithError:(NSError *)error {
    // The request has failed for some reason!
    self.statusLabel.text = @"Connection failed.";
}

- (void)receiveDidStart
{
    // Clear the current data view so that we get a nice visual cue if the receive fails.
    //self.dataReceivedTextView.text = @"";
    self.statusLabel.text = @"Receiving data.";
}


- (IBAction)clearDataView:(id)sender {
    
    self.dataReceivedTextView.text = @"";
    
    // reset the buffer
    [dataBuffer setLength:0];
    
    self.statusLabel.text = @"Data cleard...";
    self.getURLButton.enabled = TRUE;
}

- (IBAction)connectToURL:(id)sender {
    
    BOOL                success;
    NSURL *             url;
    NSURLRequest *      request;
    
   
    // First get and check the URL.
    url = [self smartURLForString:self.urlText.text];
    success = (url != nil);
   
    // If the URL is bogus, let the user know.  Otherwise kick off the connection.
    
    if ( ! success) {
        self.statusLabel.text = @"Invalid URL";
    }
    else
    {
        // Open a connection for the URL.
        request = [NSURLRequest requestWithURL:url];
          
        // Tell the UI we're receiving.
        [self receiveDidStart];
        
        dataBuffer = [NSMutableData new];
        
        //set this delegate as this object. Therefore "connection: didFailWithError:"
        //will get called if implemented
        connection = [NSURLConnection connectionWithRequest:request delegate:self];
    }
}

- (NSURL *)smartURLForString:(NSString *)str
{
    NSURL *     result;
    NSString *  trimmedStr;
    NSRange     schemeMarkerRange;
    NSString *  scheme;
    
    assert(str != nil);
    
    result = nil;
    
    trimmedStr = [str stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
    if ( (trimmedStr != nil) && (trimmedStr.length != 0) ) {
        schemeMarkerRange = [trimmedStr rangeOfString:@"://"];
        
        if (schemeMarkerRange.location == NSNotFound) {
            result = [NSURL URLWithString:[NSString stringWithFormat:@"http://%@", trimmedStr]];
        } else {
            scheme = [trimmedStr substringWithRange:NSMakeRange(0, schemeMarkerRange.location)];
            assert(scheme != nil);
            
            if ( ([scheme compare:@"http"  options:NSCaseInsensitiveSearch] == NSOrderedSame)
                || ([scheme compare:@"https" options:NSCaseInsensitiveSearch] == NSOrderedSame) ) {
                result = [NSURL URLWithString:trimmedStr];
            } else {
                // It looks like this is some unsupported URL scheme.
            }
        }
    }
    return result;
}

@end

File: 1-GettingStarted/GettingStarted-Swift/GettingStarted-Swift/HttpViewController.swift
/* Copyright (c) 2022 BlackBerry Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 */

import UIKit

class HttpViewController: UIViewController, URLSessionDelegate {
    
    var usernameText : UITextField?
    
    var passwordText : UITextField?
    
    @IBOutlet weak var getDataButton: UIButton!
    @IBOutlet weak var dataRecievedTextView: UITextView!
    
    @IBOutlet weak var urlText: UITextField!
    override func viewDidLoad() {
        super.viewDidLoad()
        
        //TODO: Start with public URL
        self.urlText.text = "http://www.example.com"
        
    }
    
    //MARK: Request Control
    
    // Requests the data from a specific URL. This can trigger authentication dialogs
    
    @IBAction func connectToUrl(_ sender: Any) {
        // First get and check the URL.
        let url = smartURLForString(str: self.urlText.text!)
        let success = (url != nil)
        
        // If the URL is bogus, let the user know.  Otherwise kick off the connection.
        if (!success) {
            
        }
        else {
            // Tell the UI we're receiving.
            
            requestData(url: self.urlText.text!, relaxSSL: false)
        }
    }
    func smartURLForString(str : String) -> NSURL? {
        
        var result : NSURL? = nil
        
        let trimmedStr = str.trimmingCharacters(in: CharacterSet.whitespaces)
        
        if trimmedStr.count != 0 {
            let schemeMarkerRange : NSRange = (trimmedStr as NSString).range(of: "://")
            if schemeMarkerRange.location == NSNotFound {
                result = NSURL(string: "http://\(trimmedStr)")!
            } else {
                let scheme = (trimmedStr as NSString).substring(with: NSRange.init(location: 0, length: schemeMarkerRange.location))
                if (scheme as NSString).compare("http", options: .caseInsensitive) == .orderedSame || (scheme as NSString).compare("https", options: .caseInsensitive) == .orderedSame {
                    result = NSURL(string: trimmedStr)!
                } else {
                    
                }
            }
        }
        return result
    }
    
    // Requests the data from a specific URL. This can trigger authentication dialogs
    
    func requestData(url : String, relaxSSL : Bool) {
        
        //Open the request using NSURLConnection or NSURLSession
        
        let nsUrl = URL(string: url)
        let urlRequest = URLRequest(url: nsUrl!)
        let session = URLSession.init(configuration: URLSessionConfiguration.default, delegate: self, delegateQueue: OperationQueue.init())
        
        let task = session.dataTask(with: urlRequest, completionHandler: { (data, response, error) in
            
            DispatchQueue.main.async {
                print(response ?? "No Response")
                print(error?.localizedDescription ?? "No Error")
                
                if data != nil {
                    self.dataRecievedTextView.text = String.init(data: data!, encoding: .utf8)
                }
                
                self.getDataButton.isEnabled = true;
            }
        })
        task.resume()
    }
    
}
